gnmi_server/server.go

```diff
@@ -515,3 +527,63 @@ func (s *Server) Capabilities(ctx context.Context, req *gnmipb.CapabilityRequest
 		GNMIVersion:        "0.7.0",
 		Extension:          exts}, nil
 }
+
+type uint128 struct {
+	High uint64
+	Low  uint64
+}
+
+func (lh *uint128) Compare(rh *uint128) int {
+	if rh == nil {
+		// For MA disabled case, EID supposed to be 0.
+		rh = &uint128{High: 0, Low: 0}
+	}
+	if lh.High > rh.High {
+		return 1
+	}
+	if lh.High < rh.High {
+		return -1
+	}
+	if lh.Low > rh.Low {
+		return 1
+	}
+	if lh.Low < rh.Low {
+		return -1
+	}
+	return 0
+}
+
+// ReqFromMasterEnabledMA returns true if the request is sent by the master
+// controller.
+func ReqFromMasterEnabledMA(req *gnmipb.SetRequest, masterEID *uint128) bool {
+	// Read the election_id.
+	reqEID := uint128{High: 0, Low: 0}
+	// It can be one of many extensions, so iterate through them to find it.
+	for _, e := range req.GetExtension() {
+		ma := e.GetMasterArbitration()
+		if ma == nil {
+			continue
+		}
+		// The Master Arbitration descriptor has been found.
+		// If MA is enabled, the reqEID needs to be set to 0
+		if ma.ElectionId != nil {
+			reqEID = uint128{High: ma.ElectionId.High, Low: ma.ElectionId.Low}
+		}
+		// Use the election ID that is in the last extension, so, no 'break' here.
+	}
+	switch masterEID.Compare(&reqEID) {
+	case 1: // This Election ID is smaller than the known Master Election ID.
+		log.V(1).Infoln("Election ID is smaller than the current master. Rejected.")
+		return false
+	case -1: // New Master Election ID received!
+		log.V(0).Infof("New master has been elected with %v\n", reqEID)
+		*masterEID = reqEID
```

Comment:
This can create a race when two set requests arrive at the same time. Better to use atomic operations

Reply:
I add mutex here to prevent racing issue.

Reply:
This local mutex will not guard against parallel set calls. It should be made a global variable.
Also, i feel the `masterEID.Compare()` call also should be guarded.

Reply:
Added mutex on EID.

Reply:
I think this will not work. It will overwrite the mutex in masterEID as well and the Unlock() call in next line will fail. Have you verified this?

Also, this code intends to lock/unlock the mutex separately for EID compare and update operations. Ideally both should be one atomic operation. Otherwise there is a chance (in theory) of EID getting overwritten by another goroutine after the compare was done but before the value is updated. Better to synchronize the entire ReqFromMasterEnabledMA logic, something like:

```go

var maMu sync.Mutex

func ReqFromMasterEnabledMA(req *gnmipb.SetRequest, masterEID *uint128) error {
    /*** find reqEID ***/
    maMu.Lock()
    defer maMu.Unlock()
    switch masterEID.Compare(reqEID) {
    case -1:
        *masterEID = reqEID
    ....
    ...
```

Reply:
Done.